pymc/func_utils.py

```diff
@@ -129,39 +141,39 @@ def find_constrained_prior(
             "need it."
         )
 
-    cdf_error = (pm.math.exp(logcdf_upper) - pm.math.exp(logcdf_lower)) - mass
-    cdf_error_fn = pm.aesaraf.compile_pymc([dist_params], cdf_error, allow_input_downcast=True)
+    target = (aet.exp(logcdf_lower) - mass_below_lower) ** 2
+    target_fn = pm.aesaraf.compile_pymc([dist_params], target, allow_input_downcast=True)
+
+    constraint = aet.exp(logcdf_upper) - aet.exp(logcdf_lower)
+    constraint_fn = pm.aesaraf.compile_pymc([dist_params], constraint, allow_input_downcast=True)
 
     jac: Union[str, Callable]
+    constraint_jac: Union[str, Callable]
     try:
-        aesara_jac = pm.gradient(cdf_error, [dist_params])
+        aesara_jac = pm.gradient(target, [dist_params])
         jac = pm.aesaraf.compile_pymc([dist_params], aesara_jac, allow_input_downcast=True)
+        aesara_constraint_jac = pm.gradient(constraint, [dist_params])
+        constraint_jac = pm.aesaraf.compile_pymc(
+            [dist_params], aesara_constraint_jac, allow_input_downcast=True
+        )
     # when PyMC cannot compute the gradient
     except (NotImplementedError, NullTypeGradError):
         jac = "2-point"
+        constraint_jac = "2-point"
+    cons = optimize.NonlinearConstraint(constraint_fn, lb=mass, ub=mass, jac=constraint_jac)
 
-    opt = optimize.least_squares(cdf_error_fn, x0=list(init_guess.values()), jac=jac)
+    opt = optimize.minimize(
+        target_fn, x0=list(init_guess.values()), jac=jac, constraints=cons, **kwargs
+    )
     if not opt.success:
-        raise ValueError("Optimization of parameters failed.")
+        raise ValueError(
+            f"Optimization of parameters failed.\nOptimization termination details:\n{opt}"
+        )
 
     # save optimal parameters
     opt_params = {
         param_name: param_value for param_name, param_value in zip(init_guess.keys(), opt.x)
     }
     if fixed_params is not None:
         opt_params.update(fixed_params)
-
-    # check mass in interval is not too far from `mass`
-    opt_dist = distribution.dist(**opt_params)
-    mass_in_interval = (
-        pm.math.exp(pm.logcdf(opt_dist, upper)) - pm.math.exp(pm.logcdf(opt_dist, lower))
-    ).eval()
-    if (np.abs(mass_in_interval - mass)) > 0.01:
-        warnings.warn(
-            f"Final optimization has {(mass_in_interval if mass_in_interval.ndim < 1 else mass_in_interval[0])* 100:.0f}% of probability mass between "
-            f"{lower} and {upper} instead of the requested {mass * 100:.0f}%.\n"
-            "You may need to use a more flexible distribution, change the fixed parameters in the "
-            "`fixed_params` dictionary, or provide better initial guesses."
-        )
-
```

Comment:
Are we sure we don't need this check anymore @lucianopaz ?

Reply:
Yes, we don't need it anymore. Now that the mass condition is an optimization constraint, the optimization will fail if the condition is not met.